### PR TITLE
Mixin & Hook for plotting hidden activations

### PR DESCRIPTION
**Pull request by karangrewal**
This introduces

- a mixin class `PlotHiddenActivations` used for plotting hidden activations of specified modules,
- a forward hook class `TrackHiddenActivationsHook` (which subclasses 'TrackStatsHookBase') for keeping track of the hidden activations computed during forward passes,
- plotting functions to visualize the activations by unit, by class; note that these plotting functions don't take 'dendrite_activations' and 'winning_mask' as input since they have nothing to do with dendrites.

The resulting visualizations looks like [this](https://drive.google.com/file/d/1kQqn6m-Gx8viF2rnHAWDfUhzl8nB6nog/view?usp=sharing), and the mixin and hook classes are designed to mimic the `PlotDendriteMetrics` and `ApplyDendritesHook` classes, respectively.

### CODE REVIEW

File: nupic/research/frameworks/dendrites/metrics/metrics.py [hunk 1]
```diff
@@ -157,6 +157,44 @@ def dendrite_activations_by_unit(dendrite_activations, winning_mask, targets):
         return selected_activations
 
 
+def hidden_activations_by_unit(activations, targets):
+    """
+    Returns a 2D torch tensor with shape (num_categories, num_units) where cell c, i
+    gives the mean value (post-sigmoid) of the selected dendrite activation for unit i
```
Comment (mvacaporale) [inline nupic/research/frameworks/dendrites/metrics/metrics.py:163]: When you say, "post-sigmoid", does this mean the input to this function comes from the output of a sigmoid?

File: nupic/research/frameworks/pytorch/hooks/track_hidden_activations.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        :param y: output of module
+        """
+        if not self._tracking:
+            return
+
+        self._activations = self._activations.to(y.device)
+        self._activations = torch.cat((y, self._activations), dim=0)
+
+        # Keep only the last 'num_samples'
+        self._activations = self._activations[:self.num_samples, ...]
```
Comment (mvacaporale) [inline nupic/research/frameworks/pytorch/hooks/track_hidden_activations.py:60]: Related to my "post-sigmoid" question, do you intend to track a `torch.nn.Sigmoid` module with this hook? I know it could work for most modules; however, the reason I ask is to make sure you're not intending to track an `ApplyDendrites` module (e.g. `DendriticGate1d`) as these return a namedtuple for which you'd have to access via `y.values`.

File: nupic/research/frameworks/vernon/mixins/__init__.py [hunk 3]
```diff
@@ -19,10 +19,15 @@
 # http://numenta.org/licenses/
 # ----------------------------------------------------------------------
 
+from .composite_loss import CompositeLoss
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/__init__.py:22]: Thanks for organizing these.

File: nupic/research/frameworks/vernon/mixins/hidden_activations.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    TrackHiddenActivationsHook,
+)
+from nupic.research.frameworks.pytorch.model_utils import filter_modules
+
+__all__ = [
+    "PlotHiddenActivations",
+]
+
+
+class PlotHiddenActivations(metaclass=abc.ABCMeta):
```
Comment (mvacaporale) [inline nupic/research/frameworks/vernon/mixins/hidden_activations.py:40]: I'm glad this pattern of hook management and plotting is useful. There's definitely some repeated logic here. We don't need to worry about this much now. If anything, it would be useful however to include a test for this mixin just in case if we try and refactor the code to reduce repeated logic.

Comment (mvacaporale) [reply nupic/research/frameworks/vernon/mixins/hidden_activations.py:40]: See `dendrite_metric_mixins_test.py` for an example test. You can even just add to that file if you'd like.

**Review by mvacaporale — APPROVED** [review]
Looks good overall. I'd urge to add a test to the new mixin before merging, however. If we keep using this paradigm, I'll probably refactor the code later on, and a test will greatly help with that.

File: nupic/research/frameworks/dendrites/metrics/metrics.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/metrics/metrics.py:163]: I think including this term is erroneous, since I mostly copied the description from `dendrite_activations_by_unit`. This has nothing to do with sigmoids, and I'll update the docstring.

The input to this function `activations` is just the output of a forward module (intended to be a dendritic layer or fc layer) over a batch of inputs. The parameter description for `activations` is correct.

File: nupic/research/frameworks/pytorch/hooks/track_hidden_activations.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/hooks/track_hidden_activations.py:60]: I don't intend to track a `torch.nn.Sigmoid` module, but it would work for that. In any case, I don't plan to use this hook for dendrites at all; it should only be used where the output is 2D, with shape (batch_size, num_features).

File: nupic/research/frameworks/vernon/mixins/hidden_activations.py (hunk 4 above)
Comment (karangrewal) [reply nupic/research/frameworks/vernon/mixins/hidden_activations.py:40]: I'll add some test cases before merging.

Comment (karangrewal) [reply nupic/research/frameworks/vernon/mixins/hidden_activations.py:40]: Test cases have been added.

File: nupic/research/frameworks/dendrites/metrics/metrics.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/dendrites/metrics/metrics.py]: Updated.
